pkg/sdkserver/sdkserver_test.go

```diff
@@ -1469,6 +1461,14 @@ func TestSDKServerAddListValue(t *testing.T) {
 			assert.Equal(t, testCase.want.Values, got.Values)
 			assert.Equal(t, testCase.want.Capacity, got.Capacity)
 
+			// start workerqueue processing at this point, so there is no chance of processing the above updates
+			// earlier.
+			sc.gsWaitForSync.Add(1)
```

Comment:
Given that the go routine was moved down, do we need the `Done()` and `Add(1)`? You already moved the `sc.Run`.

Reply:
@igooch asked me offline the same question.

So in this test we're faking things a bit, since we have this WaitGroup in the sdkserver:

https://github.com/googleforgames/agones/blob/5f58fce4c00f97ef012db8f8a503898a4725d11b/pkg/sdkserver/sdkserver.go#L242

This WaitGroup exists to pause any requests for `sdk.GameServer()` data within the SDK until the SDK has received this information from the Informer.

https://github.com/googleforgames/agones/blob/5f58fce4c00f97ef012db8f8a503898a4725d11b/pkg/sdkserver/sdkserver.go#L452

In this test, we manually start the pieces of the client-go machinery that we need (which is usually done in `sc.Run(...)` ~ line 1440, but ignore the pieces we don't want (primarily starting the workerqueue).

```
 			sc.informerFactory.Start(ctx.Done())
			require.True(t, cache.WaitForCacheSync(ctx.Done(), sc.gameServerSynced))
			sc.gsWaitForSync.Done()
```

So we're calling `sc.gsWaitForSync.Done()` because we know that the `InformerFactory` has started, we've waited for `WaitForCacheSync`, so the information is there, and `sc.gameServer()` 

But this does mean, that when we get to calling `sc.Run(...)` this will call `sc.gsWaitForSync.Done()` (as above), which will panic, as there is nothing for it to decrement on on. So we `sc.gsWaitForSync.Add(1)` to give the WaitGroup something to decrement on with `.Done()`, even though most of the work being done in `sc.Run(...)` is redundant at this point.

Does that answer the question?

Reply:
Not exactly. I understand where the existing sync points are and that there's a `gsWaitForSync.Done()` called in `sc.Run()`. But here on new line 1441 you are calling `Done()`- but there's no one that would be waiting here. You are then undoing that on new line 1466, and not starting `sc.Run()` until after that.

I'm arguing that I _think_ you can remove 1441 and 1466 and have the same guarantee - you got that guarantee by moving the `sc.Run()` lower. Unless there is something async running in new lines 1443-1462 that I'm missing?

Reply:
> but there's no one that would be waiting here

There's a .Add(1) in the inital construction of the object:

https://github.com/markmandel/agones/blob/cbbe83f7351bb9c11d83c1c98e0924bd8b41da22/pkg/sdkserver/sdkserver.go#L218

And then in this code:

https://github.com/markmandel/agones/blob/cbbe83f7351bb9c11d83c1c98e0924bd8b41da22/pkg/sdkserver/sdkserver.go#L1011-L1020

This call to `s.gameServer()` would hang without the call to `gsWaitForSync.Done()` since it has his this in it:

https://github.com/markmandel/agones/blob/cbbe83f7351bb9c11d83c1c98e0924bd8b41da22/pkg/sdkserver/sdkserver.go#L451

Reply:
Oh, I missed that `GetList` was calling `s.gameServer()` under the hood - that clarifies it. (I was misreading it as something else. :/)

Sounds good, understood!